### DISCUSSION TITLE
Function as stimuli parameter

### DISCUSSION THREAD

**Discussion by perfect-pitch-studies** [jspsych/jsPsych#193] [Q&A]
We're trying to use jsPysch to code a backward digit span task, and it seems that the single-stim plugin can accept a function as the stimuli parameter, but it doesn't do what we expect it to. Normally, when it's given an array of HTML strings directly, it displays them in sequence, but when it's given a function that returns an array, it displays everything all at once. Is this intended, and are there any workarounds that let you make stimulus lists of arbitrary length in real time? For reference, this is the relevant part of our code (the referenced global variables and functions are initialized elsewhere):

```
function getStimuli() {
  var digit;
  var stimList = [];
  currentDigitList = getDigitList(digitSpan);
  reversedDigitString = "";
  for (var i = 0; i < currentDigitList.length; i += 1) {
    digit = currentDigitList[i].toString();
    stimList.push('<p>' + digit + '</p>');
    reversedDigitString = digit + reversedDigitString;
  }
  return stimList;
}

var digitTextSeq = {
  type: "single-stim",
  stimuli: getStimuli,
  is_html: true,
  timing_response: 750,
  response_ends_trial: false,
  timing_post_trial: 250
};
```

**Comment by jodeleeuw** [comment]
The function gets evaluated at the start of each trial, which is why the behavior isn't what you were expecting.

For the code below, supplying an array to the stimuli parameter generates 4 distinct trials with different numbers.

``` javascript
var trials = {
  type: 'single-stim',
  stimuli: ['1','2','3','4'],
  is_html: true
}
```

But for this declaration:

``` javascript
var trials = {
  type: 'single-stim',
  stimuli: function(){ return ['1','2','3','4'] },
  is_html: true
}
```

jsPsych interprets this as a **single** trial, because there is only one item passed to the stimuli parameter and it is not an array. So when the trial begins, the function is evaluated, and the stimulus is 4 different numbers.

A workaround is a bit complicated, because arbitrary list lengths mean that the number of trials is not known in advance. So you'll have to use a loop. Here's a sketch of how that might work. (Disclaimer: I'm generating this without testing -- it may not be bug free)

``` javascript
var digit_index = 0;
var digits = ['1','2','3','4'];

var trial = {
  type: 'single-stim',
  stimuli: function(){
    var digit = digits[digit_index];
    digit_index++;
    return('<p>'+digit+'</p>');
  },
  is_html: true
}

var loop = {
  chunk_type: 'while',
  timeline: [trial],
  continue_function: function(){
    if(digit_index >= digits.length){
       return false;
    } else {
       return true;
    }
}

jsPsych.init({
  experiment_structure: [loop]
});
```

**Comment by perfect-pitch-studies** [comment]
Thanks for the explanation! That makes sense.

I'm not sure using while chunks would necessarily work for us since we also want to be able to terminate the experiment early based on the accuracy of the participant's answers. We figured it would be reasonable enough to pre-generate random arrays of digits up to a length of 15 and push conditional chunks to the experiment structure array.

So far, this is doing what we want it to:

```
for (var i = 2; i < 16; i += 1) {
  var stimSeq;
  var questions;
  var stimList = getStimuli(i);

  if (useAudio) {
    stimSeq = {
      type: "single-audio",
      stimuli: stimList,
      timing_response: 1000,
      response_ends_trial: false,
      timing_post_trial: 0
    };
    questions = [['Please type the digits you just heard in reverse order (starting from the last digit and ending on the first digit).']];
  } else {
    stimSeq = {
      type: "single-stim",
      stimuli: stimList,
      is_html: true,
      timing_response: 900,
      response_ends_trial: false,
      timing_post_trial: 100
    };
    questions = [['Please type the digits you just saw in reverse order (starting from the last digit and ending on the first digit).']];
  }

  var response = {
    type: "survey-text",
    questions: questions,
    data: [{
      correct_answer: reversedDigitString
    }],
    on_finish: function (trialData) {
      var answer = JSON.parse(trialData.responses).Q0;
      wasCorrect = (answer === trialData.correct_answer);
      jsPsych.data.addDataToLastTrial({
        answer: answer,
        digit_span: (wasCorrect ? digitSpan: (digitSpan - 1)),
        was_correct: (wasCorrect ? 1 : 0)
      });
      if (wasCorrect) {
        digitSpan += 1;
      }
    },
    timing_post_trial: 0
  };

  var ifChunk = {
    chunk_type: "if",
    timeline: [fixation, stimSeq, response],
    conditional_function: function () {
      return wasCorrect;
    }
  };

  experiment.push(ifChunk);
}
```

**Comment by jodeleeuw** [comment]
Glad it is working!
